File: binding/src/libqwineventnotifier.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qwineventnotifier.html
pub const qwineventnotifier = struct {

    /// New constructs a new QWinEventNotifier object.
    pub fn New() QtC.QWinEventNotifier {
        return qtc.QWinEventNotifier_new();
    }


    /// New2 constructs a new QWinEventNotifier object.
    pub fn New2(hEvent: ?*anyopaque) QtC.QWinEventNotifier {
        return qtc.QWinEventNotifier_new2(@ptrCast(hEvent));
    }


    /// New3 constructs a new QWinEventNotifier object.
    pub fn New3(parent: ?*anyopaque) QtC.QWinEventNotifier {
        return qtc.QWinEventNotifier_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QWinEventNotifier object.
    pub fn New4(hEvent: ?*anyopaque, parent: ?*anyopaque) QtC.QWinEventNotifier {
        return qtc.QWinEventNotifier_new4(@ptrCast(hEvent), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWinEventNotifier_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwineventnotifier.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetHandle(self: ?*anyopaque, hEvent: ?*anyopaque) void {
        qtc.QWinEventNotifier_SetHandle(@ptrCast(self), @ptrCast(hEvent));
    }

    pub fn Handle(self: ?*anyopaque, ) ?*anyopaque {
        return qtc.QWinEventNotifier_Handle(@ptrCast(self));
    }

    pub fn IsEnabled(self: ?*anyopaque, ) bool {
        return qtc.QWinEventNotifier_IsEnabled(@ptrCast(self));
    }

    pub fn SetEnabled(self: ?*anyopaque, enable: bool) void {
        qtc.QWinEventNotifier_SetEnabled(@ptrCast(self), enable);
    }

    pub fn Event(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QWinEventNotifier_Event(@ptrCast(self), @ptrCast(e));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QWinEventNotifier_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QWinEventNotifier_QBaseEvent(@ptrCast(self), @ptrCast(e));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWinEventNotifier_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwineventnotifier.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWinEventNotifier_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwineventnotifier.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Wrapper to allow calling private signal
    pub fn Onactivated(self: ?*anyopaque, callback: *const fn (?*anyopaque, HANDLE) callconv(.c) void) void {
        qtc.QWinEventNotifier_Connect_activated(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWinEventNotifier_Delete(@ptrCast(self));
    }
};

File: binding/src/libqwineventnotifier.cpp
#include <QEvent>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWinEventNotifier>
#include <qwineventnotifier.h>
#include "libqwineventnotifier.h"
#include "libqwineventnotifier.hxx"

QWinEventNotifier* QWinEventNotifier_new() {
	 return new VirtualQWinEventNotifier();
}

QWinEventNotifier* QWinEventNotifier_new2(void** hEvent) {
	 return new VirtualQWinEventNotifier(hEvent);
}

QWinEventNotifier* QWinEventNotifier_new3(QObject* parent) {
	 return new VirtualQWinEventNotifier(parent);
}

QWinEventNotifier* QWinEventNotifier_new4(void** hEvent, QObject* parent) {
	 return new VirtualQWinEventNotifier(hEvent, parent);
}

libqt_string QWinEventNotifier_Tr(const char* s) {
	QString _ret = QWinEventNotifier::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QWinEventNotifier_SetHandle(QWinEventNotifier* self, void** hEvent) {
	self->setHandle(hEvent);
}

void** QWinEventNotifier_Handle(const QWinEventNotifier* self) {
	return self->handle();
}

bool QWinEventNotifier_IsEnabled(const QWinEventNotifier* self) {
	return self->isEnabled();
}

void QWinEventNotifier_SetEnabled(QWinEventNotifier* self, bool enable) {
	self->setEnabled(enable);
}

libqt_string QWinEventNotifier_Tr2(const char* s, const char* c) {
	QString _ret = QWinEventNotifier::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QWinEventNotifier_Tr3(const char* s, const char* c, int n) {
	QString _ret = QWinEventNotifier::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Derived class handler implementation
bool QWinEventNotifier_Event(QWinEventNotifier* self, QEvent* e) {
	auto* vqwineventnotifier = dynamic_cast<VirtualQWinEventNotifier*>(self);
	if (vqwineventnotifier && vqwineventnotifier->isVirtualQWinEventNotifier) {
	return vqwineventnotifier->event(e);
	} else {
	return self->QWinEventNotifier::event(e);
}
}

// Base class handler implementation
bool QWinEventNotifier_QBaseEvent(QWinEventNotifier* self, QEvent* e) {
	auto* vqwineventnotifier = dynamic_cast<VirtualQWinEventNotifier*>(self);
	if (vqwineventnotifier && vqwineventnotifier->isVirtualQWinEventNotifier) {
vqwineventnotifier->setQWinEventNotifier_Event_IsBase(true);
	return vqwineventnotifier->event(e);
}
}

// Auxiliary method to allow providing re-implementation
void QWinEventNotifier_OnEvent(QWinEventNotifier* self, intptr_t slot) {
	auto* vqwineventnotifier = dynamic_cast<VirtualQWinEventNotifier*>(self);
	if (vqwineventnotifier && vqwineventnotifier->isVirtualQWinEventNotifier) {
vqwineventnotifier->setQWinEventNotifier_Event_Callback(reinterpret_cast<VirtualQWinEventNotifier::QWinEventNotifier_Event_Callback>(slot));
}
}

void QWinEventNotifier_Connect_activated(QWinEventNotifier* self, intptr_t slot) {
    void (*slotFunc)(QWinEventNotifier*, HANDLE) = reinterpret_cast<void (*)(QWinEventNotifier*, HANDLE)>(slot);
    QWinEventNotifier::connect(self, &QWinEventNotifier::activated, [self, slotFunc](HANDLE hEvent) {
	slotFunc(self, hEvent);
    });
}

void QWinEventNotifier_Delete(QWinEventNotifier* self) {
    delete self;
}

